src/System.Management.Automation/engine/Subsystem/Prediction/IPredictor.cs

```diff
@@ -0,0 +1,163 @@
+
+        /// <summary>
+        /// A command line was accepted to execute.
+        /// The predictor can start processing early as needed with the latest history.
+        /// </summary>
+        /// <param name="history">History command lines provided as references for prediction.</param>
+        void StartEarlyProcessing(IReadOnlyList<string> history);
+
+        /// <summary>
+        /// The suggestion given by the predictor was accepted.
+        /// </summary>
+        /// <param name="acceptedSuggestion">The accepted suggestion text.</param>
+        void OnSuggestionAccepted(string acceptedSuggestion);
+
+        /// <summary>
+        /// Get the predictive suggestions.
+        /// </summary>
+        /// <param name="context">The <see cref="PredictionContext"/> object to be used for prediction.</param>
+        /// <param name="cancellationToken">The cancellation token to cancel the prediction.</param>
+        /// <returns>A list of predictive suggestions.</returns>
+        List<PredictiveSuggestion>? GetSuggestion(PredictionContext context, CancellationToken cancellationToken);
+    }
+
+    /// <summary>
+    /// Context information about the user input.
+    /// </summary>
+    public class PredictionContext
+    {
+        /// <summary>
+        /// Gets the abstract syntax tree (AST) generated from parsing the user input.
+        /// </summary>
+        public Ast InputAst { get; }
+
+        /// <summary>
+        /// Gets the tokens generated from parsing the user input.
+        /// </summary>
+        public IReadOnlyList<Token> InputTokens { get; }
+
+        /// <summary>
+        /// Gets the cursor position, which is assumed always at the end of the input line.
+        /// </summary>
+        public IScriptPosition CursorPosition { get; }
+
+        /// <summary>
+        /// Gets the token at the cursor.
+        /// </summary>
+        public Token? TokenAtCursor { get; }
+
+        /// <summary>
+        /// Gets all ASTs that are related to the cursor position,
+        /// which is assumed always at the end of the input line.
+        /// </summary>
+        public IReadOnlyList<Ast> RelatedAsts { get; }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="PredictionContext"/> class from the AST and tokens that represent the user input.
+        /// </summary>
+        /// <param name="inputAst">The <see cref="Ast"/> object from parsing the current command line input.</param>
+        /// <param name="inputTokens">The <see cref="Token"/> objects from parsing the current command line input.</param>
+        public PredictionContext(Ast inputAst, Token[] inputTokens)
+        {
+            Requires.NotNull(inputAst, nameof(inputAst));
+            Requires.NotNull(inputTokens, nameof(inputTokens));
+
+            var cursor = inputAst.Extent.EndScriptPosition;
+            var astContext = CompletionAnalysis.ExtractAstContext(inputAst, inputTokens, cursor);
```

Comment:
I don't think it's necessary to do this in all places where `var` is used.
Take this one as an example: the type is `CompletionAnalysis.AstAnalysisContext`, a nested type. It's long and doesn't have additional value to your readability because it doesn't give more information than the name of the variable, and you will have to look into the code of `ExtractAstContext` and the type itself to learn more.